File: app/src/main/java/fr/xgouchet/chronorg/data/models/Jump.java
package fr.xgouchet.chronorg.data.models;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import org.joda.time.DateTime;
import org.joda.time.ReadableInstant;

/**
 * @author Xavier Gouchet
 */
public class Jump implements Parcelable {

    private int id;
    private int entityId;
    private int order;
    @NonNull private String name = "";
    @NonNull private ReadableInstant from = new DateTime("1985-10-26T01:35:00-08:00");
    @NonNull private ReadableInstant to = new DateTime("1955-11-05T06:15:00-08:00");

    public Jump() {
    }

    public Jump(@NonNull ReadableInstant from, @NonNull ReadableInstant to) {
        this.from = from;
        this.to = to;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    public void setEntityId(int entityId) {
        this.entityId = entityId;
    }

    public int getEntityId() {
        return entityId;
    }

    public void setName(@NonNull String name) {
        this.name = name;
    }

    @NonNull public String getName() {
        return name;
    }

    public void setFrom(@NonNull ReadableInstant from) {
        this.from = from;
    }

    public void setFrom(@NonNull String from) {
        this.from = new DateTime(from);
    }

    @NonNull public ReadableInstant getFrom() {
        return from;
    }

    public void setTo(@NonNull ReadableInstant to) {
        this.to = to;
    }

    public void setTo(@NonNull String to) {
        this.to = new DateTime(to);
    }

    @NonNull public ReadableInstant getTo() {
        return to;
    }

    public void setOrder(int order) {
        this.order = order;
    }

    public int getOrder() {
        return order;
    }

    @Override public int describeContents() {
        return 0;
    }

    @Override public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(this.id);
        dest.writeInt(this.entityId);
        dest.writeInt(this.order);
        dest.writeString(this.name);
        dest.writeString(this.from.toString());
        dest.writeString(this.to.toString());
    }

    protected Jump(Parcel in) {
        this.id = in.readInt();
        this.entityId = in.readInt();
        this.order = in.readInt();
        this.name = in.readString();
        this.from = new DateTime(in.readString());
        this.to = new DateTime(in.readString());
    }

    public static final Parcelable.Creator<Jump> CREATOR = new Parcelable.Creator<Jump>() {
        @Override public Jump createFromParcel(Parcel source) {
            return new Jump(source);
        }

        @Override public Jump[] newArray(int size) {
            return new Jump[size];
        }
    };
}

File: khronorg/src/main/java/fr/xgouchet/khronorg/feature/jumps/Jump.kt
package fr.xgouchet.khronorg.feature.jumps

import android.os.Parcel
import android.os.Parcelable
import fr.xgouchet.khronorg.commons.time.getLocalTimeZone
import fr.xgouchet.khronorg.data.models.Direction
import org.joda.time.DateTime
import org.joda.time.Interval
import org.joda.time.ReadableInstant
import org.joda.time.ReadableInterval

/**
 * @author Xavier F. Gouchet
 */
data class Jump(var id: Int = -1,
                var travellerId: Int = -1,
                var order: Int = 0,
                var from: ReadableInstant = DateTime("1985-10-26T01:35:00-08:00"),
                var delay: ReadableInterval = Interval(DateTime("1955-11-05T06:15:00-08:00"), DateTime("1985-10-26T01:35:00-08:00")),
                var direction: Int = Direction.PAST)
    : Parcelable {

    constructor(id: Int = -1,
                travellerId: Int = -1,
                order: Int = 0,
                from: ReadableInstant = DateTime("1985-10-26T01:35:00-08:00"),
                dest: ReadableInstant = DateTime("1955-11-05T06:15:00-08:00"))
            : this(id, travellerId, order, from,
            if (from.isBefore(dest)) Interval(from, dest) else Interval(dest, from),
            if (from.isBefore(dest)) Direction.FUTURE else Direction.PAST)

    val destination: ReadableInstant
        get() {
            when (direction) {
                Direction.PAST -> return from.toInstant().minus(delay.toDuration())
                Direction.FUTURE -> return from.toInstant().plus(delay.toDuration())
                else -> throw IllegalArgumentException("Unknown jump direction : $direction")
            }
        }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        dest?.let {
            it.writeInt(id)
            it.writeInt(travellerId)
            it.writeInt(order)
            it.writeString(from.toString())
            it.writeString(delay.toString())
            it.writeInt(direction)
        }
    }

    constructor(input: Parcel) : this() {
        id = input.readInt()
        travellerId = input.readInt()
        order = input.readInt()
        from = DateTime(input.readString(), getLocalTimeZone())
        delay = Interval(input.readString())
        direction = input.readInt()
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<Jump> = object : Parcelable.Creator<Jump> {
            override fun createFromParcel(source: Parcel): Jump {
                return Jump(source)
            }

            override fun newArray(size: Int): Array<Jump?> {
                return arrayOfNulls(size)
            }
        }

        val ID_BIRTH = -1;
        val ID_DEATH = -2;
    }
}